app/javascript/components/retirement-form/index.jsx

```diff
@@ -0,0 +1,88 @@
+import React, { useEffect, useState } from 'react';
+
+import MiqFormRenderer from '@@ddf';
+import createSchema from './retirement-form.schema';
+import { API, http } from '../../http_api';
+import handleFailure from '../../helpers/handle-failure';
+
+const RetirementForm = ({ retirementID }) => {
+  console.log(retirementID)
+  const id = retirementID.split('"').filter(Number);
+  const [{ initialValues, isLoading }, setState] = useState({
+    isLoading: !!id,
+    initialValues: {
+      retirementDate: null,
+      retirementWarning: '',
+      formMode: '',
+    },
+  });
+
+  console.log(`retirementID ${id} ${typeof id}`);
+  const cancelUrl = `/${ManageIQ.controller}/retire?button=cancel`;
+  const saveURL = `/${ManageIQ.controller}/retire?button=save`;
+
+  const calculateDate = (months, weeks, days, hours) => {
+    const now = new Date();
+    now.setDate(now.getDate() + (weeks * 7) + days);
+    now.setMonth(now.getMonth() + months);
+    now.setHours(now.getHours() + hours);
+    return now;
+  };
+
+  const onSubmit = (data) => {
+    miqSparkleOn();
+    const date = data.formMode === 'delay' ? calculateDate(parseInt(data.months, 10), parseInt(data.weeks, 10), parseInt(data.days, 10), parseInt(data.hours, 10)) : data.retirementDate;
+    const request = data.formMode === 'delay'
+      ? API.post(`/api/services/${id}`, { action: 'request_retire', resource: { date, warn: data.retirementWarning } })
+      : API.post(`/api/services/${id}`, { action: 'request_retire', resource: { date, warn: data.retirementWarning } });
+    request.then(() => {
+      miqAjaxButton(saveURL, { retire_date: date, retire_warn: data.retirementWarning });
+    }).catch(miqSparkleOff);
+  };
+
+  useEffect(() => {
+    if (id.length === 1) {
+      http.get(`/${ManageIQ.controller}/retirement_info/${id}`).then((res) => {
```

Comment:
The API should be able to provide us this information, if you query for the entity. Look into the definition of the `retirement_info` method in the Ruby codebase to find out what kind of objects have retirement attributes.

For services I was able to fetch:
```js 
# curl -XGET "http://admin:smartvm@localhost:3000/api/services/141?attributes=retires_on,retirement_warn" | jq
{
  "href": "http://localhost:3000/api/services/141",
  "retires_on": "2018-11-21T14:42:01Z",
  "retirement_warn": null,
  "id": "141"
}
```